### PR TITLE
Extend numpy.size to accept multiple axes.

### PR DESCRIPTION
**Pull request by carlosgmartin**
Extends [`jax.numpy.size`](https://docs.jax.dev/en/latest/_autosummary/jax.numpy.size.html) to accept multiple axes.

For context:
- https://github.com/numpy/numpy/issues/29239
- https://github.com/numpy/numpy/pull/29240

### CODE REVIEW

File: jax/_src/numpy/util.py [hunk 1]
```diff
@@ -453,6 +453,10 @@ def size(a: ArrayLike | SupportsSize | SupportsShape, axis: int | None = None) -
     6
     >>> jnp.size(y, axis=1)
     3
+    >>> jnp.size(y, axis=(1,))
+    3
+    >>> jnp.size(y, axis=(0, 1))
+    6
```
Comment (jakevdp) [inline jax/_src/numpy/util.py:460]: Did the updated implementation not make it into the commit?

Comment (carlosgmartin) [reply jax/_src/numpy/util.py:460]: About that: The [current implementation](https://github.com/jax-ml/jax/blob/7106f757bb04cfe23f4fb034652a76d38564bfe1/jax/_src/numpy/util.py#L471) of `jax.numpy.size` dispatches to `np.size`. Does that mean that, when the next NumPy is released (which will handle a tuple `axis`), JAX will also handle a tuple `axis` as-is?

Comment (jakevdp) [reply jax/_src/numpy/util.py:460]: Exactly: since JAX delegates to NumPy, it means that JAX returns whatever is returned by the version of NumPy that you have installed. If we want JAX to support multiple axes in `jnp.size`, JAX will have to re-implement the `size` function until NumPy 2.4.0 is the oldest supported version.

File: jax/_src/numpy/util.py [hunk 2]
```diff
@@ -464,12 +470,13 @@ def size(a: ArrayLike | SupportsSize | SupportsShape, axis: int | None = None) -
...
-  if (axis is None and hasattr(a, "size")) or (axis is not None and hasattr(a, "shape")):
-    # NumPy dispatches to a.size/a.shape if available.
-    return np.size(a, axis=axis)  # type: ignore[arg-type]
-  # Deprecation warning added 2025-2-20.
   check_arraylike("size", a, emit_warning=True)
-  if hasattr(a, "__jax_array__"):
-    a = a.__jax_array__()
-  # NumPy dispatches to a.size/a.shape if available.
-  return np.size(a, axis=axis)  # type: ignore[arg-type]
+  _shape = asarray(a).shape
```
Comment (jakevdp) [inline jax/_src/numpy/util.py]: The current implementation dispatches to `a.size` or `a.shape` if defined: unconditionally passing the input to `asarray` will cause breakages (e.g. slowdowns due to unexpected data transfer, or errors when `a` cannot be converted to a JAX array or if transfer guard is enabled).

We should continue to dispatch to `a.size` and/or `a.shape` when it's available. Basically this implementation should more-or-less be a direct backport of NumPy's new implementation, with a TODO to remove it once the minimal supported version contains the new logic.

Comment (carlosgmartin) [reply jax/_src/numpy/util.py]: Ok, I'll dispatch to `a.size` or `a.shape` when available.

Out of curiosity, what's an example of an `a` that can be passed to `size()` but not `asarray()`?

And what's an example of a `SupportsSize` that's not also a `SupportsShape`?

Comment (jakevdp) [reply jax/_src/numpy/util.py]: e.g. something like this currently works:
```python
class Foo:
  size = 4
jnp.size(Foo())
```
Do people actually rely on this? I've been doing this long enough that I no longer assume the answer is "no" 😁

Comment (carlosgmartin) [reply jax/_src/numpy/util.py]: Ah, I was wondering if there was a more "concrete" / "array-like" example. I was curious as to why `size()` would be designed to be so generic.

Comment (carlosgmartin) [reply jax/_src/numpy/util.py]: After all, NumPy's own [size](https://github.com/numpy/numpy/blob/82bd6ef31ce57859280be7591c3a20f562e69d34/numpy/_core/fromnumeric.pyi#L1400) is designed to handle only `ArrayLike`.

Comment (carlosgmartin) [reply jax/_src/numpy/util.py]: I've updated the implementation.

However, long-term, I think it would be a good idea to narrow the API of the `size`, `shape`, and `ndim` functions to accept only `ArrayLike`, like NumPy does.

Comment (jakevdp) [reply jax/_src/numpy/util.py]: JAX's `size` works this way because NumPy's `size` works this way.

> After all, NumPy's own [size](https://github.com/numpy/numpy/blob/82bd6ef31ce57859280be7591c3a20f562e69d34/numpy/_core/fromnumeric.pyi#L1400) is designed to handle only `ArrayLike`.

That's just a type annotation. The runtime implementation allows arbitrary inputs with a `size` or `shape` attribute.

Comment (jakevdp) [reply jax/_src/numpy/util.py]: A more concrete example of where this is important: say someone has a custom array implementation that uses `__jax_array__` to convert to JAX arrays. The current implementation of `jnp.size` will return the size without incurring the cost of conversion to a JAX array, which may involve actually moving bytes to the device – I think it's important to maintain that behavior.

**Review by jakevdp — CHANGES_REQUESTED** [review]
Looks good – can we add some tests? Maybe update `testSizeAlongAxis` to test tuples as well as integers

File: jax/_src/numpy/util.py [hunk 3]
```diff
@@ -464,12 +469,20 @@ def size(a: ArrayLike | SupportsSize | SupportsShape, axis: int | None = None) -
...
+  else:
+    if hasattr(a, "shape"):
+      _shape = a.shape
+    else:
+      _shape = np.shape(a)  # type: ignore[arg-type]
+
+    if isinstance(axis, int):
+      return _shape[axis]
+    elif isinstance(axis, tuple):
+      return math.prod(_shape[ax] for ax in axis)
```
Comment (jakevdp) [inline jax/_src/numpy/util.py]: What happens if there are duplicates in the `axis` list? Does the result here match the NumPy behavior?

(it would be nice to have unit tests against NumPy that cover these kinds of corner cases, skipped for older versions that don't support a tuple axis)

Comment (carlosgmartin) [reply jax/_src/numpy/util.py]: Does JAX have something analogous to NumPy's [`normalize_axis_tuple`](https://numpy.org/doc/stable/reference/generated/numpy.lib.array_utils.normalize_axis_tuple.html)? NumPy's `size` function is implemented [with it](https://github.com/numpy/numpy/blob/82bd6ef31ce57859280be7591c3a20f562e69d34/numpy/_core/fromnumeric.py#L3613).

Comment (jakevdp) [reply jax/_src/numpy/util.py]: There's `jax._src.util.canonicalize_axis`

File: jax/_src/numpy/reductions.py [hunk 4]
```diff
@@ -784,17 +784,6 @@ def amax(a: ArrayLike, axis: Axis = None, out: None = None,
...
 
-def _axis_size(a: ArrayLike, axis: int | Sequence[int]):
-  if not isinstance(axis, (tuple, list)):
-    axis_seq: Sequence[int] = (axis,)  # type: ignore[assignment]
-  else:
-    axis_seq = axis
-  size = 1
-  a_shape = np.shape(a)
-  for a in axis_seq:
-    size *= maybe_named_axis(a, lambda i: a_shape[i], lax_parallel.axis_size)
```
Comment (jakevdp) [inline jax/_src/numpy/reductions.py]: I'm worried by the fact that replacing this utility with your `size` function removes handling of named axes.

One option would be to revert this part of the change – another option would be to add named axis handling to the new `jnp.size`. What do you think?

Comment (carlosgmartin) [reply jax/_src/numpy/reductions.py]: [`_axis_size`](https://github.com/jax-ml/jax/blob/6f8456e9fbf776dc85f5d7cf9b0bade9c658a76c/jax/_src/numpy/reductions.py#L787) only accepts `axis: int | Sequence[int]`. So it wasn't handling named axes anyway. And it was only used in two places ([here](https://github.com/jax-ml/jax/blob/6f8456e9fbf776dc85f5d7cf9b0bade9c658a76c/jax/_src/numpy/reductions.py#L892) and [here](https://github.com/jax-ml/jax/blob/6f8456e9fbf776dc85f5d7cf9b0bade9c658a76c/jax/_src/numpy/reductions.py#L1130)) with the same argument type `axis: int | Sequence[int]`.

That said, if you want to preserve it, perhaps because it might be useful in the future for handling named axes, I can revert that part of the change. Let me know.

> another option would be to add named axis handling to the new `jnp.size`

I don't think NumPy's [`numpy.size`](https://numpy.org/doc/stable/reference/generated/numpy.size.html) handles named axes, though. Wouldn't that break API equivalence?

Comment (jakevdp) [reply jax/_src/numpy/reductions.py]: type annotations are different than runtime types: in general, just because something is annotated one way does not mean those are the only types passed to the function at runtime. We could run the tests to see exactly what this breaks, but I'm fairly confident it will break something (Hyrum's law)

Comment (jakevdp) [reply jax/_src/numpy/reductions.py]: > Wouldn't that break API equivalence?

There is no problem here: in general, we can't guarantee API equivalence between `jax.numpy` and `numpy`. For example, `jax.numpy` operations usually return JAX arrays, while `numpy` equivalents usually return NumPy arrays 😁

Comment (carlosgmartin) [reply jax/_src/numpy/reductions.py]: I've extended `size` to accept named axes.

File: jax/_src/numpy/util.py [hunk 5]
```diff
@@ -464,12 +470,28 @@ def size(a: ArrayLike | SupportsSize | SupportsShape, axis: int | None = None) -
...
+
+  if axis is None:
+    return math.prod(_shape)
+
+  elif isinstance(axis, int):
+    return maybe_named_axis(axis, lambda i: _shape[i], lax_parallel.axis_size)
+
+  elif isinstance(axis, Sequence):
+    ndim = len(_shape)
+    axis = tuple(maybe_named_axis(i, lambda i: canonicalize_axis(i, ndim), lambda name: name) for i in axis)
```
Comment (jakevdp) [inline jax/_src/numpy/util.py]: I think we should use `lax_parallel.axis_size` here, and then remove the second call to `maybe_named_axis` below. The reason is that each axis may be referred to by a name or a number, and so the `len(axis) != len(set(axis))` check below may miss collisions in the current version. What do you think?

Comment (jakevdp) [reply jax/_src/numpy/util.py]: Wait a second, I'm wrong: we don't want axis size here, we want axis index!

I wonder if there's a way to convert axis name to axis index?

File: jax/_src/numpy/util.py (hunk 2 above)
Comment (carlosgmartin) [reply jax/_src/numpy/util.py]: That makes me wonder: Is there an analogue of this for `dtype`? I often do `jnp.asarray(x).dtype` just to get the dtype if it exists, but as you point out, that [might be wasteful](https://github.com/google-deepmind/optax/pull/1375#issuecomment-3075480326).

It looks like I'm [not alone](https://github.com/search?q=jax+%2Fasarray%5C%28.*%5C%29%5C.dtype%2F&type=code).

Unfortunately, [`numpy.dtype`](https://docs.jax.dev/en/latest/_autosummary/jax.numpy.dtype.html) already means something else, breaking symmetry with [`numpy.shape`](https://docs.jax.dev/en/latest/_autosummary/jax.numpy.shape.html) and [`numpy.ndim`](https://docs.jax.dev/en/latest/_autosummary/jax.numpy.ndim.html).

File: jax/_src/numpy/util.py (hunk 5 above)
Comment (carlosgmartin) [reply jax/_src/numpy/util.py]: I'm not sure. Are "named axes" and axis indices even comparable/inter-convertible (i.e. do they have a one-to-one correspondence)?

If you prefer, I can leave the original `_axis_size` and its users untouched, and limit this PR to just changing `numpy.size` itself, for the time being.

File: jax/_src/numpy/util.py (hunk 2 above)
Comment (jakevdp) [reply jax/_src/numpy/util.py]: Yeah dtypes are a whole bag of worms. I don't know of any builtin API that does "get the dtype attribute if it exists, and otherwise convert to array and then return the dtype".

All that aside though, if we want to land this change to `jnp.size`, we will need to do it in a backward-compatible way to prevent breaking existing users of the API.

File: jax/_src/numpy/util.py (hunk 5 above)
Comment (jakevdp) [reply jax/_src/numpy/util.py]: The simplest thing to do here would be to forget about named axes entirely: the current version of `jnp.size` does not touch them, so expanding its scope seems ill-advised unless there's good reason to do so.

Comment (carlosgmartin) [reply jax/_src/numpy/util.py]: To clarify: Do you want me to leave the original `_axis_size` and its users untouched, and edit only `jnp.size`? Or do you want me to use `jnp.size` for the users and get rid of `_axis_size`?

Comment (jakevdp) [reply jax/_src/numpy/util.py]: I would probably leave `_axis_size` untouched, and just change `jnp.size` to have the `axis` argument without support for named axes. That's probably the simplest way to land this change, and it will let us go back to basically delegating to `np.size` once the `axis` argument is supported there.

Comment (carlosgmartin) [reply jax/_src/numpy/util.py]: Done.

**Review by jakevdp — APPROVED** [review]
Thanks!

File: jax/_src/numpy/util.py [hunk 6]
```diff
@@ -464,12 +469,23 @@ def size(a: ArrayLike | SupportsSize | SupportsShape, axis: int | None = None) -
...
-  return np.size(a, axis=axis)  # type: ignore[arg-type]
+  if axis is None and hasattr(a, "size"):
+    return a.size
+
+  _shape = shape(a)  # type: ignore[arg-type]
+
+  if axis is None:
+    return math.prod(_shape)
+
+  elif isinstance(axis, int):
```
Comment (jakevdp) [inline jax/_src/numpy/util.py]: A potential regression here: I believe `jnp.size(x, np.int32(0))` would error after this change.

Comment (carlosgmartin) [reply jax/_src/numpy/util.py]: Should I make it `int | np.ndarray | jax.Array`?

Comment (jakevdp) [reply jax/_src/numpy/util.py]: I don't care about type annotations, I care about runtime regressions. We should fix the `if /elif` statement to treat inputs the same way `np.size` does.

Comment (carlosgmartin) [reply jax/_src/numpy/util.py]: I meant for the runtime `isinstance` check.

I made the latest version match [NumPy's implementation](https://github.com/numpy/numpy/blob/778647f32ba37d6f8c8cc5772a4a0734b715e785/numpy/_core/fromnumeric.py#L3605).

File: jax/_src/util.py [hunk 7]
```diff
@@ -405,6 +405,17 @@ def canonicalize_axis(axis: SupportsIndex, num_dims: int) -> int:
...
+def canonicalize_axis_tuple(axis, ndim, allow_duplicate=False):
+  if axis is None:
+    return tuple(range(ndim))
+  if isinstance(axis, Sequence):
+    axis = tuple(canonicalize_axis(i, ndim) for i in axis)
+    if not allow_duplicate and len(set(axis)) != len(axis):
+      raise ValueError(f"repeated axis: {axis}")
+    return axis
+  else:
+    return (axis,)
```
Comment (jakevdp) [inline jax/_src/util.py]:
```suggestion
    return (canonicalize_axis(axis, ndim),)
```

File: jax/_src/util.py [hunk 8]
```diff
@@ -405,6 +405,17 @@ def canonicalize_axis(axis: SupportsIndex, num_dims: int) -> int:
     axis = axis + num_dims
   return axis
 
+def canonicalize_axis_tuple(axis, ndim, allow_duplicate=False):
```
Comment (jakevdp) [inline jax/_src/util.py]: Add annotations?

**Review by jakevdp — APPROVED** [review]
Looks good! Thanks for putting up with all the back-and-forth during review

File: jax/_src/numpy/util.py [hunk 9]
```diff
@@ -464,12 +469,8 @@ def size(a: ArrayLike | SupportsSize | SupportsShape, axis: int | None = None) -
     >>> y.size
     6
   """
-  if (axis is None and hasattr(a, "size")) or (axis is not None and hasattr(a, "shape")):
-    # NumPy dispatches to a.size/a.shape if available.
-    return np.size(a, axis=axis)  # type: ignore[arg-type]
-  # Deprecation warning added 2025-2-20.
-  check_arraylike("size", a, emit_warning=True)
```
Comment (jakevdp) [inline jax/_src/numpy/util.py:471]: The `testNdimShapeSizeNonArrayInput` test  expects the warning generated by this check. I think restoring this line is probably the best bet.

File: jax/_src/numpy/util.py [hunk 10]
```diff
@@ -464,12 +469,9 @@ def size(a: ArrayLike | SupportsSize | SupportsShape, axis: int | None = None) -
     >>> y.size
     6
   """
-  if (axis is None and hasattr(a, "size")) or (axis is not None and hasattr(a, "shape")):
-    # NumPy dispatches to a.size/a.shape if available.
-    return np.size(a, axis=axis)  # type: ignore[arg-type]
-  # Deprecation warning added 2025-2-20.
+  if axis is None and hasattr(a, "size"):
+    return a.size
   check_arraylike("size", a, emit_warning=True)
```
Comment (jakevdp) [inline jax/_src/numpy/util.py]: Please put this above the previous check. We should require array-like input regardless of whether `axis` is None.

**Review by jakevdp — APPROVED** [review]
